# Import Libraries

In [1]:
import numpy as np
import pandas as pd 
import warnings 
warnings.filterwarnings("ignore")

# Read CSV File

In [2]:
df = pd.read_csv('customer.csv')
df.head()

,ID_code,target,var_0,var_1,var_2,var_3,var_4,var_5,var_6,var_7,...,var_190,var_191,var_192,var_193,var_194,var_195,var_196,var_197,var_198,var_199
0,train_0,0,8.9255,-6.7863,11.9081,5.0930,11.4607,-9.2834,5.1187,18.6266,...,4.4354,3.9642,3.1364,1.6910,18.5227,-2.3978,7.8784,8.5635,12.7803,-1.0914
1,train_1,0,11.5006,-4.1473,13.8588,5.3890,12.3622,7.0433,5.6208,16.5338,...,7.6421,7.7214,2.5837,10.9516,15.4305,2.0339,8.1267,8.7889,18.3560,1.9518
2,train_2,0,8.6093,-2.7457,12.0805,7.8928,10.5825,-9.0837,6.9427,14.6155,...,2.9057,9.7905,1.6704,1.6858,21.6042,3.1417,-6.5213,8.2675,14.7222,0.3965
3,train_3,0,11.0604,-2.1518,8.9522,7.1957,12.5846,-1.8361,5.8428,14.9250,...,4.4666,4.7433,0.7178,1.4214,23.0347,-1.2706,-2.9275,10.2922,17.9697,-8.9996
4,train_4,0,9.8369,-1.4834,12.8746,6.6375,12.2772,2.4486,5.9405,19.2514,...,-1.4905,9.5214,-0.1508,9.1942,13.2876,-1.5121,3.9267,9.5031,17.9974,-8.8104


It is Huge Dataset, There are a total of 200 features in this data set along with ID_code and target columns. The target columns contain 0 and 1 value where 0 means the customer will not do a transaction and 1 means the customer will do a transaction. It has Target column which helps to the model that customer has Transaction.

# Basic Checks

In [3]:
df.shape

(200000, 202)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Columns: 202 entries, ID_code to var_199
dtypes: float64(200), int64(1), object(1)
memory usage: 308.2+ MB


In [5]:
df['target'].value_counts()

target
0    179902
1     20098
Name: count, dtype: int64

Checked data shape, types, and imbalance in target variable.Identified that dataset is large and highly imbalanced (9:1).This has to be Handled correctly.

In [6]:
df.isnull().sum()

ID_code    0
target     0
var_0      0
var_1      0
var_2      0
          ..
var_195    0
var_196    0
var_197    0
var_198    0
var_199    0
Length: 202, dtype: int64

There no such NULL values in the Dataset

In [7]:
df.describe()

,target,var_0,var_1,var_2,var_3,var_4,var_5,var_6,var_7,var_8,...,var_190,var_191,var_192,var_193,var_194,var_195,var_196,var_197,var_198,var_199
count,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,...,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000,200000.000000
mean,0.100490,10.679914,-1.627622,10.715192,6.796529,11.078333,-5.065317,5.408949,16.545850,0.284162,...,3.234440,7.438408,1.927839,3.331774,17.993784,-0.142088,2.303335,8.908158,15.870720,-3.326537
std,0.300653,3.040051,4.050044,2.640894,2.043319,1.623150,7.863267,0.866607,3.418076,3.332634,...,4.559922,3.023272,1.478423,3.992030,3.135162,1.429372,5.454369,0.921625,3.010945,10.438015
min,0.000000,0.408400,-15.043400,2.117100,-0.040200,5.074800,-32.562600,2.347300,5.349700,-10.505500,...,-14.093300,-2.691700,-3.814500,-11.783400,8.694400,-5.261000,-14.209600,5.960600,6.299300,-38.852800
25%,0.000000,8.453850,-4.740025,8.722475,5.254075,9.883175,-11.200350,4.767700,13.943800,-2.317800,...,-0.058825,5.157400,0.889775,0.584600,15.629800,-1.170700,-1.946925,8.252800,13.829700,-11.208475
50%,0.000000,10.524750,-1.608050,10.580000,6.825000,11.108250,-4.833150,5.385100,16.456800,0.393700,...,3.203600,7.347750,1.901300,3.396350,17.957950,-0.172700,2.408900,8.888200,15.934050,-2.819550
75%,0.000000,12.758200,1.358625,12.516700,8.324100,12.261125,0.924800,6.003000,19.102900,2.937900,...,6.406200,9.512525,2.949500,6.205800,20.396525,0.829600,6.556725,9.593300,18.064725,4.836800
max,1.000000,20.315000,10.376800,19.353000,13.188300,16.671400,17.251600,8.447700,27.691800,10.151300,...,18.440900,16.716500,8.402400,18.281800,27.928800,4.272900,18.321500,12.000400,26.079100,28.500700


# Preprocessing

In [8]:
df.drop('ID_code',axis=1,inplace=True)           #Unique Feature
df.head()

,target,var_0,var_1,var_2,var_3,var_4,var_5,var_6,var_7,var_8,...,var_190,var_191,var_192,var_193,var_194,var_195,var_196,var_197,var_198,var_199
0,0,8.9255,-6.7863,11.9081,5.0930,11.4607,-9.2834,5.1187,18.6266,-4.9200,...,4.4354,3.9642,3.1364,1.6910,18.5227,-2.3978,7.8784,8.5635,12.7803,-1.0914
1,0,11.5006,-4.1473,13.8588,5.3890,12.3622,7.0433,5.6208,16.5338,3.1468,...,7.6421,7.7214,2.5837,10.9516,15.4305,2.0339,8.1267,8.7889,18.3560,1.9518
2,0,8.6093,-2.7457,12.0805,7.8928,10.5825,-9.0837,6.9427,14.6155,-4.9193,...,2.9057,9.7905,1.6704,1.6858,21.6042,3.1417,-6.5213,8.2675,14.7222,0.3965
3,0,11.0604,-2.1518,8.9522,7.1957,12.5846,-1.8361,5.8428,14.9250,-5.8609,...,4.4666,4.7433,0.7178,1.4214,23.0347,-1.2706,-2.9275,10.2922,17.9697,-8.9996
4,0,9.8369,-1.4834,12.8746,6.6375,12.2772,2.4486,5.9405,19.2514,6.2654,...,-1.4905,9.5214,-0.1508,9.1942,13.2876,-1.5121,3.9267,9.5031,17.9974,-8.8104


Removed unnecessary columns like ID_code is necessary beacause it has unique Values which doesn't give information to predict and ensured clean, usable data.Prepared data for modeling without missing values or noise.

# Feature & Target Split

In [9]:
x = df.drop('target',axis=1)
y = df['target']

In [10]:
x.shape

(200000, 200)

In [11]:
y.shape

(200000,)

### Feature Engineering

In [12]:
#These new features help model understand:
x["mean"] = x.mean(axis=1)                 #Overall level (mean, sum)
x["std"] = x.std(axis=1)                    #Variation (std)
x["min"] = x.min(axis=1)                  #Extremes (min, max)
x["max"] = x.max(axis=1)
x["sum"] = x.sum(axis=1)

In [13]:
x["pos_count"] = (x > 0).sum(axis=1)           #POsitive Counts
x["neg_count"] = (x < 0).sum(axis=1)               #Negative Counts
x["range"] = x["max"] - x["min"]           #Range from Mx to Min

Created statistical features (Mean, std, Min, Max, Negative,Range etc.) to capture hidden patterns. These features improved model learning and boosted performance. Which the Input has not more information to predict.

# Train Test Split

In [14]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,stratify=y,random_state=42)

Used stratified split to maintain class distribution in both sets. Ensured reliable and unbiased model evaluation. To control from imbalance we used stratify for target column.

# Preprocessing

In [15]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

To get data into one singe scale which avoids overfitting and all postive and negative values will scaled.

# Model Building

In [16]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(class_weight="balanced",C=0.01, max_iter=1000)     #Handle Imbalance

In [17]:
lr = model.fit(x_train,y_train)
lr

LogisticRegression(C=0.01, class_weight='balanced', max_iter=1000)

# Prediction

In [18]:
y_predict = lr.predict(x_test)
y_predict

array([0, 0, 0, ..., 0, 1, 0])

In [19]:
x_test.shape

(40000, 208)

In [20]:
y_test.shape

(40000,)

# Evaluation

In [25]:
from sklearn.metrics import roc_auc_score
lr_c = roc_auc_score(y_test, y_predict)
lr_c

np.float64(0.787211386093435)

In [24]:
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_test, y_predict))

[[28535  7445]
 [  879  3141]]


In [26]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_predict))

              precision    recall  f1-score   support

           0       0.97      0.79      0.87     35980
           1       0.30      0.78      0.43      4020

    accuracy                           0.79     40000
   macro avg       0.63      0.79      0.65     40000
weighted avg       0.90      0.79      0.83     40000



Used ROC-AUC as main metric due to imbalance. Also analyzed confusion matrix, precision, and recall.With this model ROC-AUC score is 78. As we see the confusion matrix that model has predicting only 0 the Customer who has transacted is not predicting .The model precision score is 30 which is very low.But FP is very low which is good.

# XGBoost Model

In [27]:
from xgboost import XGBClassifier
model = XGBClassifier(
     n_estimators=100, 
    max_depth=5,
    learning_rate=0.1,
    tree_method="hist",                #Uses histogram bins
    scale_pos_weight=9             #Handle Imbalance
)

In [28]:
model.fit(x_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              feature_weights=None, gamma=None, grow_policy=None,
              importance_type=None, interaction_constraints=None,
              learning_rate=0.1, max_bin=None, max_cat_threshold=None,
              max_cat_to_onehot=None, max_delta_step=None, max_depth=5,
              max_leaves=None, min_child_weight=None, missing=nan,
              monotone_constraints=None, multi_strategy=None, n_estimators=100,
              n_jobs=None, num_parallel_tree=None, ...)

# Prediction

In [29]:
y_pred = model.predict_proba(x_test)[:,1]          #Selects only probability of class 1

# Threshold Tuning

In [30]:
from sklearn.metrics import confusion_matrix
for t in [0.3, 0.4, 0.5, 0.6, 0.7,0.8,0.9]:
    y_predict = (y_pred > t).astype(int)
    print(f"Threshold: {t}")
    print(confusion_matrix(y_test, y_predict))

Threshold: 0.3
[[13728 22252]
 [  170  3850]]
Threshold: 0.4
[[23546 12434]
 [  586  3434]]
Threshold: 0.5
[[30224  5756]
 [ 1277  2743]]
Threshold: 0.6
[[33832  2148]
 [ 2135  1885]]
Threshold: 0.7
[[35424   556]
 [ 2964  1056]]
Threshold: 0.8
[[35907    73]
 [ 3641   379]]
Threshold: 0.9
[[35978     2]
 [ 3984    36]]


In [31]:
y_final= (y_pred > 0.6).astype(int)

Adjusted prediction threshold to balance precision and recall. Improved control over false positives and false negatives. As the confusion matrix shows the threshold if we keep low the precision is getting high. If threshold is more than 0.6 the Recall is getting high. Then when it has 0.6 than it has balanced between Precision and Recall.

# Evaluate

In [32]:
from sklearn.metrics import roc_auc_score
roc = roc_auc_score(y_test, y_pred)
roc

np.float64(0.84960755215031)

In [33]:
print("Logistic:", lr_c)
print("XGBoost:", roc)

Logistic: 0.787211386093435
XGBoost: 0.84960755215031


In [34]:
print(confusion_matrix(y_test, y_final))


[[33832  2148]
 [ 2135  1885]]


In [35]:
print(classification_report(y_test, y_final))

              precision    recall  f1-score   support

           0       0.94      0.94      0.94     35980
           1       0.47      0.47      0.47      4020

    accuracy                           0.89     40000
   macro avg       0.70      0.70      0.70     40000
weighted avg       0.89      0.89      0.89     40000



The model achieved a ROC-AUC score of ~0.85 and accuracy of ~89%, indicating strong ability to distinguish between classes despite imbalance. XGBoost outperformed Logistic Regression by capturing complex patterns in the data.

Through feature engineering and imbalance handling, the model achieved a good balance between precision and recall, making it reliable for predicting customer transactions. Overall, the model is robust, efficient, and suitable for real-world classification tasks.

# Problems faced 

1. Imbalanced Dataset

The dataset had a high imbalance (9:1 ratio), making the model biased towards the majority class. This affected metrics like accuracy and required special handling techniques.

2.  Computational Limitations

Large dataset size (200k rows) caused slow training and system crashes during GridSearchCV. Required reducing model complexity and using efficient methods.

3. Feature Engineering Challenges

Identifying useful features without domain knowledge was difficult. Trial-and-error was required to find features that improved performance.

4. Precision-Recall Trade-off

Improving recall led to lower precision and vice versa. Finding the right balance using threshold tuning was challenging.